pkg/controllers/node/termination/controller.go

```diff
@@ -138,6 +152,56 @@ func (c *Controller) removeFinalizer(ctx context.Context, n *v1.Node) error {
 	return nil
 }
 
+func (c *Controller) terminationGracePeriodExpirationTime(ctx context.Context, node *v1.Node) (*time.Time, error) {
+
+	nodeClaim := &v1beta1.NodeClaim{}
+
+	if len(node.OwnerReferences) == 0 {
+		logging.FromContext(ctx).Errorf("node has no owner, could not find NodeClaim for Node: %v", node.Name)
+		return nil, nil
+	}
+
+	nodeClaimName := types.NamespacedName{}
+	for _, o := range node.OwnerReferences {
+		// verify that we're finding Karpenter's NodeClaim of any version permutation
+		if strings.HasPrefix(o.APIVersion, "karpenter.sh") && o.Kind == "NodeClaim" {
+			nodeClaimName = types.NamespacedName{
+				Name: o.Name,
+			}
+		}
+	}
+
+	if err := c.kubeClient.Get(ctx, nodeClaimName, nodeClaim); err != nil {
+		return nil, err
+	}
+
+	// TODO: remove, holding this here in case we switch the preferred implementation
+	// nodePool := &v1beta1.NodePool{}
+	// // assume the only NodeClaim ownerRef is the NodePool
+	// nodePoolName := types.NamespacedName{
+	// 	Name: nodeClaim.OwnerReferences[0].Name,
+	// }
+	// if err := c.kubeClient.Get(ctx, nodePoolName, nodePool); err != nil {
+	// 	logging.FromContext(ctx).Errorf("could not find NodePool for NodeClaim: %v", nodeClaim.Name)
+	// 	return nil, err
+	// }
+
+	// if nodePool.Spec.Disruption.TerminationGracePeriod != nil {
+	// 	expirationTime := node.DeletionTimestamp.Time.Add(nodePool.Spec.Disruption.TerminationGracePeriod.Duration)
+	// 	c.recorder.Publish(terminatorevents.NodeTerminationGracePeriod(node, expirationTime, fmt.Sprintf("%s", nodePool.Spec.Disruption.TerminationGracePeriod)))
+	// 	// logging.FromContext(ctx).Infof("node %v will be forcefully terminated at %v (terminationGracePeriod=%v)", node.Name, expirationTime, nodePool.Spec.Disruption.TerminationGracePeriod)
+	// 	return &expirationTime, nil
+	// }
+
+	if nodeClaim.Spec.TerminationGracePeriod != nil {
+		expirationTime := node.DeletionTimestamp.Time.Add(nodeClaim.Spec.TerminationGracePeriod.Duration)
```

Comment:
This is a tiny difference most likely but is the terminationGracePeriod here based on the deletion timestamp of the node or the deletion timestamp of the NodeClaim? To me -- I was thinking the NodeClaim since the terminationGracePeriod is a property of the NodeClaim

Reply:
That's a good catch, I missed this.